pennylane/pauli/conversion.py

```diff
@@ -0,0 +1,140 @@
+        hide_identity (bool): does not include the Identity observable within
+            the tensor products of the decomposition if ``True``.
+        wire_order (list[Union[int, str]]): the ordered list of wires with respect
+            to which the operator is represented as a matrix.
+        pauli (bool): return a PauliSentence instance if ``True``.
+
+    Returns:
+        Union[~.Hamiltonian, ~.PauliSentence]: the matrix decomposed as a linear combination
+            of Pauli operators, either as a ``~.Hamiltonian`` or ``~.PauliSentence`` instance.
+
+    **Example:**
+
+    We can use this function to compute the Pauli operator decomposition of an arbitrary Hermitian
+    matrix:
+
+    >>> A = np.array(
+    ... [[-2, -2+1j, -2, -2], [-2-1j,  0,  0, -1], [-2,  0, -2, -1], [-2, -1, -1,  0]])
+    >>> H = pauli_decompose(A)
+    >>> print(H)
+    (-1.0) [I0 I1]
+    + (-1.5) [X1]
+    + (-0.5) [Y1]
+    + (-1.0) [Z1]
+    + (-1.5) [X0]
+    + (-1.0) [X0 X1]
+    + (-0.5) [X0 Z1]
+    + (1.0) [Y0 Y1]
+    + (-0.5) [Z0 X1]
+    + (-0.5) [Z0 Y1]
+
+    We can return a ``~.PauliSentence`` instance by using the keyword argument ``pauli=True``:
+
+    >>> ps = pauli_decompose(A, pauli=True)
+    >>> print(ps)
+    -1.0 * I
+    + -1.5 * X(1)
+    + -0.5 * Y(1)
+    + -1.0 * Z(1)
+    + -1.5 * X(0)
+    + -1.0 * X(0) @ X(1)
+    + -0.5 * X(0) @ Z(1)
+    + 1.0 * Y(0) @ Y(1)
+    + -0.5 * Z(0) @ X(1)
+    + -0.5 * Z(0) @ Y(1)
+
+    We can also set custom wires using the ``wire_order`` argument:
+
+    >>> ps = pauli_decompose(A, pauli=True, wire_order=['a', 'b'])
+    >>> print(ps)
+    -1.0 * I
+    + -1.5 * X(b)
+    + -0.5 * Y(b)
+    + -1.0 * Z(b)
+    + -1.5 * X(a)
+    + -1.0 * X(a) @ X(b)
+    + -0.5 * X(a) @ Z(b)
+    + 1.0 * Y(a) @ Y(b)
+    + -0.5 * Z(a) @ X(b)
+    + -0.5 * Z(a) @ Y(b)
+    """
+    n = int(np.log2(len(H)))
+    N = 2**n
+
+    if wire_order is not None and len(wire_order) != n:
+        raise ValueError(
+            f"number of wires {len(wire_order)} is not compatible with number of qubits {n}"
+        )
+
+    if wire_order is None:
+        wire_order = range(n)
+
+    if H.shape != (N, N):
+        raise ValueError("The matrix should have shape (2**n, 2**n), for any qubit number n>=1")
+
+    if not np.allclose(H, H.conj().T):
+        raise ValueError("The matrix is not Hermitian")
+
+    obs_lst = []
+    coeffs = []
+
+    for term in product([I, X, Y, Z], repeat=n):
+        matrices = [mat_map[i] for i in term]
+        coeff = np.trace(reduce(np.kron, matrices) @ H) / N
+        coeff = np.real_if_close(coeff).item()
+
+        if not np.allclose(coeff, 0):
+            obs_term = (
+                [(o, w) for w, o in zip(wire_order, term) if o != I]
+                if hide_identity
+                else [(o, w) for w, o in zip(wire_order, term)]
+            )
+
+            if obs_term:
+                coeffs.append(coeff)
+                obs_lst.append(obs_term)
+
+    if pauli:
```

Comment:
i would rather make the logic until here a private (?) function that returns `coeffs, obs_lst`, then make two more functions. something like `matrix_to_hamiltonian` and `matrix_to_pauli_sentence`, but of course names are subject to change. The mixed return type configurable by keyword-arg is not my fav style choice

Reply:
I think we are sticking with design decision of the top level `pauli_decompose` method